src/arch/riscv/tlb.cc

```diff
@@ -366,6 +368,32 @@ TLB::translate(const RequestPtr &req, ThreadContext *tc,
 
         if (!delayed && fault == NoFault) {
             pma->check(req);
+
+            // not sure if this is the
+            // best place for pmp checks
+
+            if (pmode != PrivilegeMode::PRV_M){
+                bool pass = pmp->pmp_check(req, mode);
+                // what else to do here?
+                // raise some kind of an exception
+
+                if (!pass)
+                {
+
+                    ExceptionCode code;
+                    if (mode == TLB::Read)
+                        code = ExceptionCode::LOAD_ACCESS;
+                    else if (mode == TLB::Write)
+                        code = ExceptionCode::STORE_ACCESS;
+                    else
+                        code = ExceptionCode::INST_ACCESS;
+                    fault = std::make_shared<AddressFault>
+                        (req->getVaddr(), code);
+
+                }
+
+            }
```

Comment:
So, this code is duplicated in the TLB and in the PTW. This is what I'm talking about when I say it would be good to unify the permission checking. This is OK for now, but should be done in a follow on patch.

Reply:
Yeah, that still needs to be done.